File: backend/app/schemas/analytics.py
from pydantic import BaseModel, Field


class AnalyticsScopeOut(BaseModel):
    role_view: str
    department_id: int | None = None
    user_id: int | None = None


class SummaryCardsOut(BaseModel):
    total_letters: int = 0
    pending_approval: int = 0
    under_department_processing: int = 0
    consultant_active_tasks: int = 0
    waiting_final_closure: int = 0
    officially_closed: int = 0
    rejected_letters: int = 0
    returned_for_correction: int = 0


class DepartmentAnalyticsItemOut(BaseModel):
    department_id: int
    department_name: str
    department_code: str
    total_letters: int = 0
    pending_count: int = 0
    closed_count: int = 0
    overdue_assignments: int = 0
    avg_resolution_days: float | None = None


class DepartmentAnalyticsOut(BaseModel):
    items: list[DepartmentAnalyticsItemOut] = Field(default_factory=list)
    total: int = 0
    limit: int | None = None
    offset: int | None = None


class ConsultantAnalyticsItemOut(BaseModel):
    consultant_id: int
    consultant_name: str
    consultant_email: str
    assigned_count: int = 0
    resolved_count: int = 0
    transferred_count: int = 0
    active_workload: int = 0
    overdue_tasks: int = 0
    avg_completion_days: float | None = None


class ConsultantAnalyticsOut(BaseModel):
    items: list[ConsultantAnalyticsItemOut] = Field(default_factory=list)
    top_performers: list[ConsultantAnalyticsItemOut] = Field(default_factory=list)
    overloaded_consultants: list[ConsultantAnalyticsItemOut] = Field(default_factory=list)
    total: int = 0
    limit: int | None = None
    offset: int | None = None


class WorkflowStatusItemOut(BaseModel):
    key: str
    label: str
    count: int


class WorkflowStatusOut(BaseModel):
    items: list[WorkflowStatusItemOut] = Field(default_factory=list)


class TrendPointOut(BaseModel):
    period: str
    received: int = 0
    closed: int = 0


class DepartmentTrendPointOut(BaseModel):
    period: str
    department_code: str
    letters: int = 0


class ConsultantTrendPointOut(BaseModel):
    period: str
    consultant_id: int
    consultant_name: str
    assignments: int = 0


class TrendsOut(BaseModel):
    letters: list[TrendPointOut] = Field(default_factory=list)
    departments: list[DepartmentTrendPointOut] = Field(default_factory=list)
    consultants: list[ConsultantTrendPointOut] = Field(default_factory=list)


class BottleneckItemOut(BaseModel):
    letter_id: int
    serial_no: str
    subject: str
    department_code: str | None = None
    days_pending: int


class BottleneckOut(BaseModel):
    overdue_assignments: int = 0
    delayed_closures: int = 0
    high_backlog_departments: list[DepartmentAnalyticsItemOut] = Field(default_factory=list)
    longest_pending_letters: list[BottleneckItemOut] = Field(default_factory=list)


class AnalyticsOverviewOut(BaseModel):
    scope: AnalyticsScopeOut
    summary: SummaryCardsOut
    workflow_status: WorkflowStatusOut
    bottlenecks: BottleneckOut

File: frontend/src/types/analytics.ts
export type AnalyticsScope = {
  role_view: string;
  department_id: number | null;
  user_id: number | null;
};

export type SummaryCards = {
  total_letters: number;
  pending_approval: number;
  under_department_processing: number;
  consultant_active_tasks: number;
  waiting_final_closure: number;
  officially_closed: number;
  rejected_letters: number;
  returned_for_correction: number;
};

export type WorkflowStatusItem = {
  key: string;
  label: string;
  count: number;
};

export type DepartmentAnalyticsItem = {
  department_id: number;
  department_name: string;
  department_code: string;
  total_letters: number;
  pending_count: number;
  closed_count: number;
  overdue_assignments: number;
  avg_resolution_days: number | null;
};

export type ConsultantAnalyticsItem = {
  consultant_id: number;
  consultant_name: string;
  consultant_email: string;
  assigned_count: number;
  resolved_count: number;
  transferred_count: number;
  active_workload: number;
  overdue_tasks: number;
  avg_completion_days: number | null;
};

export type TrendPoint = {
  period: string;
  received: number;
  closed: number;
};

export type DepartmentTrendPoint = {
  period: string;
  department_code: string;
  letters: number;
};

export type ConsultantTrendPoint = {
  period: string;
  consultant_id: number;
  consultant_name: string;
  assignments: number;
};

export type BottleneckItem = {
  letter_id: number;
  serial_no: string;
  subject: string;
  department_code: string | null;
  days_pending: number;
};

export type AnalyticsOverview = {
  scope: AnalyticsScope;
  summary: SummaryCards;
  workflow_status: { items: WorkflowStatusItem[] };
  bottlenecks: {
    overdue_assignments: number;
    delayed_closures: number;
    high_backlog_departments: DepartmentAnalyticsItem[];
    longest_pending_letters: BottleneckItem[];
  };
};

export type AnalyticsTablePaging = {
  limit: number;
  offset: number;
  sort_by: string;
  sort_dir: "asc" | "desc";
};

export type DepartmentAnalyticsOut = {
  items: DepartmentAnalyticsItem[];
  total: number;
  limit: number | null;
  offset: number | null;
};

export type ConsultantAnalyticsOut = {
  items: ConsultantAnalyticsItem[];
  top_performers: ConsultantAnalyticsItem[];
  overloaded_consultants: ConsultantAnalyticsItem[];
  total: number;
  limit: number | null;
  offset: number | null;
};
export type TrendsOut = {
  letters: TrendPoint[];
  departments: DepartmentTrendPoint[];
  consultants: ConsultantTrendPoint[];
};

export type AnalyticsFilters = {
  preset?: string;
  date_from?: string;
  date_to?: string;
  department_id?: number;
};
